### PR TITLE
Add vanilla-JS Shadow Reader demo app

### PR DESCRIPTION
**Pull request by pbakaus**
Yes, this is a pretty huge PR. It includes the entire vanilla-JS PWAMP demo app (deployed at https://amp.cards) that I showcased at Google I/O.

Should work fine in Chrome and Safari, Edge support and detailed write up coming soon.

### CODE REVIEW

File: amp-pwa-reader/src/js/Article.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+          reject('Article does not have an AMP version.');
+        }
+      }; // .responseXML contains a ready-to-use Document object
+      xhr.send();
+    });
+
+  }
+
+  load() {
+    return (this.doc ? new Promise(resolve => resolve()) : this.fetch().then(doc => {
```
Comment (andreban) [inline amp-pwa-reader/src/js/Article.js]: Could be written as:
`return (this.doc ? Promise.resolve() : this.fetch().then(doc => {`

File: amp-pwa-reader/package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "ShadowReader",
+  "version": "0.0.1",
+  "description": "",
+  "main": "gulpfile.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "Paul Bakaus",
+  "license": "MIT",
```
Comment (andreban) [inline amp-pwa-reader/package.json]: Should this be `Apache-2.0`?

File: amp-pwa-reader/src/js/Article.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    return new Promise((resolve, reject) => {
+      xhr.open('GET', 'https://seed-octagon.glitch.me/' + encodeURIComponent(this.url), true);
+      xhr.responseType = 'document';
+      xhr.setRequestHeader('Accept', 'text/html');
+      xhr.onload = () => {
+        var isAMP = xhr.responseXML.documentElement.hasAttribute('amp') || xhr.responseXML.documentElement.hasAttribute('⚡');
+        if (isAMP) {
+          resolve(xhr.responseXML);
+        } else {
+          reject('Article does not have an AMP version.');
```
Comment (andreban) [inline amp-pwa-reader/src/js/Article.js]: nit: Could use an early return.

File: amp-pwa-reader/src/js/Article.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    let htmlStyles = window.getComputedStyle(document.querySelector("html"));
+    this._animationSpeed = parseFloat(htmlStyles.getPropertyValue("--animation-speed")) * 1000;
+    return this._animationSpeed;
+  }
+
+  animateIn() {
+
+    // No animation if there's no card to animate from
+    if (!this.card) {
+      return new Promise(resolve => resolve());
```
Comment (andreban) [inline amp-pwa-reader/src/js/Article.js]: Can be written as:

`return Promise.resolve();`

File: amp-pwa-reader/src/js/Article.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  show(replaceHistoryState) {
+
+    // We need to clone the featured image into the Shadow DOM so it scrolls
+    // along. There are cases were we don't have a linked card from the list
+    // view (e.g. we load directly into the article), in which case we need to
+    // generate a new one.
+    var card = this.card ? this.cloneCard() : this.generateCard();
+    this.ampDoc.ampdoc.getBody().prepend(card);
+
+    return new Promise(resolve => {
```
Comment (andreban) [inline amp-pwa-reader/src/js/Article.js]: Can this simplified with the below?

     return this.animateIn().then(() => {

         // Hide the original card, show the cloned one (this also animates)
         if (this.card) {
           card.style.opacity = '1';
         }
 
         // add class to html element for to contain the scroll, and transform
         // the hamburger into a 'back' button.
         document.documentElement.classList.add('sr-article-shown');
 
         this.takeoverScroll();
 
         // Set the visibility state of the AMP doc to visible
         this.ampDoc.setVisibilityState('visible');
 
         // Finally, add new history entry
         // Note: We're doing this deliberately late due to an AMP
         // Bug that overrides the history state object early on
         shadowReader.nav.setOpenArticle(this, replaceHistoryState);
         return true;
       });

File: amp-pwa-reader/src/js/Article.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+      this.restoreScroll();
+
+      // Show the card header, hide the cloned one
+      if (this.card) {
+        this.clonedCard.style.opacity = '0';
+      }
+
+      // animate everything back to the card/listing view, then
+      // clear the old Shadow DOM to free up memory.
+      return new Promise(resolve => {
```
Comment (andreban) [inline amp-pwa-reader/src/js/Article.js]: Can this be simplified as:

    return this.animateOut().then(() => {
       this.clear();
       return true;   
    }

**Comment by andreban** [comment]
Dropped a few notes, mostly nits. The license headers seem to be missing in many files.

**Comment by pbakaus** [comment]
Thanks @andreban! Addressed all nits.

**Comment by pbakaus** [comment]
Ping! @andreban @choumx

**Pull request merged**